src/theory/arith/rewriter/addition.cpp

```diff
@@ -185,16 +186,16 @@ void addToSum(Sum& sum, TNode n, bool negate)
     multiplicity = Integer(-1);
   }
   addToProduct(monomial, multiplicity, n);
-  addToSum(sum, mkNonlinearMult(monomial), multiplicity);
+  addToSum(sum, mkNonlinearMult(monomial, n.getType()), multiplicity);
 }
 
-Node collectSum(const Sum& sum)
+Node collectSum(const Sum& sum, const TypeNode& tn)
 {
-  if (sum.empty()) return mkConst(Rational(0));
+  if (sum.empty()) return mkConst(Rational(0), tn);
   Trace("arith-rewriter") << "Collecting sum " << sum << std::endl;
   // construct the sum as nodes.
   NodeBuilder nb(Kind::ADD);
-  for (const auto& s : sum)
+  for (const std::pair<const Node, RealAlgebraicNumber>& s : sum)
```

Comment:
I'd rather keep `auto`... :)

Reply:
This use of auto makes the code hard to parse:  note `mkMultTerm` is overloaded. One cannot see what variant is being called unless the type is made explicit.

I would suggest either this change or we should not use name overloading on `mkMultTerm`, e.g. `mkMultTermRan`, `mkMultRational`.  However, I prefer having the type in the `for`.